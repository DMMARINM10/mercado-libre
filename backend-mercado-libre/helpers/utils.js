const numberSeparation = (number) => {
    const integer = Math.floor(number);
    const decimals = Number((number - integer).toFixed(2));
    return {
        integer,
        decimals,
    }
}

const arrayCategories = (data) => {
    const availableFilters = data.available_filters.find((filter) => filter.id === 'category');
    if(availableFilters) {
        const categories = availableFilters?.values?.sort((a, b) => b.results - a.results).map((value) => value.name);
        return categories;
    }
    const filter = data?.filters?.find((filter) => filter?.id === 'category');
    const categories = filter?.values[0]?.path_from_root?.map((cat) => cat.name)
    return categories;
}

module.exports = {
    numberSeparation,
    arrayCategories,
}