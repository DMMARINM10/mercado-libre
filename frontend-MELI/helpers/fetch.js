export const fetcher = (url) => {
    return fetch(url)
    .then((res) => {
        if(!res.ok) {
            throw new Error()
        }
        return res.json()
    });
}

export const getItem = async (id) => {
    const url = `${process.env.NEXT_PUBLIC_API_URL}/${id}`
    try {
        const resp = await fetch(url)
        if(!resp.ok) {
            return {
                error: true,
                data: null,
            }
        }
        const data = await resp.json()
        return {
            data,
            error: false
        }
    } catch (error) {
        console.error(error)
        return {
            error: true,
            data: null,
        }
    }
}