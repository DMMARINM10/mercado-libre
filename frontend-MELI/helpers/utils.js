export const formatNumber = (number) => {
    return number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ".");
}

export const decimalFormat = (number) => {
    const decimalString = number?.toString()
    const result = decimalString?.includes('.') 
        ? decimalString.substring(2, decimalString.length) 
        : '00'
    return result?.length !== 2 ? result + '0' : result
}