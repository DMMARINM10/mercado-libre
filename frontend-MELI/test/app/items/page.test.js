import { useSearchParams } from 'next/navigation';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import useSWR from 'swr';
import Items from '@/app/items/page';
import { formatNumber } from '@/helpers/utils';

jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn(),
}));

jest.mock('swr', () => jest.fn());

describe('Items Page Component', () => {
  beforeEach(() => {
      jest.clearAllMocks();
      useSearchParams.mockReturnValue({
          get: () => 'test-search',
        });
    });
    
    it('renders nothing when search param is empty string', () => {
        useSearchParams.mockReturnValue({
            get: () => '',
        });
        useSWR.mockReturnValue({
            data: null,
            error: null,
            isLoading: null,
        });
        
        render(<Items />);
        
        expect(screen.queryByTestId('loader')).not.toBeInTheDocument();
        expect(screen.queryByTestId('error')).not.toBeInTheDocument();
        expect(screen.queryByTestId('item-page-data')).not.toBeInTheDocument();
  });

  it('renders Loader when loading', () => {
    useSWR.mockReturnValue({
      data: null,
      error: null,
      isLoading: true,
    });

    render(<Items />);

    const loader = screen.getByTestId('loader');
    expect(loader).toBeInTheDocument();
  });

  it('renders Error when there is an error', () => {
    useSWR.mockReturnValue({
      data: null,
      error: true,
      isLoading: false,
    });

    render(<Items />);

    expect(screen.getByTestId('error')).toBeInTheDocument();
  });

  it('renders items when data is available', () => {
    const mockData = {
          categories: ['cat1', 'cat2'],
          items: [
                      {
                          id: '1',
                          picture: 'http://picture1.jpg',
                          price: { amount: 100 },
                          title: 'item1',
                          seller: 'seller1',
                          free_shipping: true,
                      },
                      {
                          id: '2',
                          picture: 'http://picture2.jpg',
                          price: { amount: 200 },
                          title: 'item2',
                          seller: 'seller2',
                          free_shipping: false,
                      },
                ],
            }
    useSWR.mockReturnValue({
      data: mockData,
      error: null,
      isLoading: false,
    });

    render(<Items />);

    const dataContainer = screen.getByTestId('items-page-data');
    expect(dataContainer).toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    useSWR.mockReturnValue({
        data: {
          categories: ['cat1', 'cat2'],
          items: [
                      {
                          id: '1',
                          picture: 'http://picture1.jpg',
                          price: { amount: 100 },
                          title: 'item1',
                          seller: 'seller1',
                          free_shipping: true,
                      },
                      {
                          id: '2',
                          picture: 'http://picture2.jpg',
                          price: { amount: 200 },
                          title: 'item2',
                          seller: 'seller2',
                          free_shipping: false,
                      },
                  ],
              },
        error: null,
        isLoading: false,
      });
    const { asFragment } = render(<Items />);
    expect(asFragment()).toMatchSnapshot();
  });
});
