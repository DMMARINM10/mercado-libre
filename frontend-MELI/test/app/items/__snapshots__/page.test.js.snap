// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items Page Component should match the snapshot 1`] = `
<DocumentFragment>
  <nav>
    <ul
      class="breadcrumb"
    >
      <li>
        cat1
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSize14px css-821wyw-MuiSvgIcon-root"
          data-testid="ArrowForwardIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
          />
        </svg>
      </li>
      <li>
        cat2
      </li>
    </ul>
  </nav>
  <section
    class="container"
    data-testid="items-page-data"
  >
    <a
      class="item-card"
      href="/items/1"
    >
      <img
        alt="picture-item-1"
        class="item-picture"
        data-nimg="1"
        decoding="async"
        height="180"
        loading="lazy"
        src="/_next/image?url=http%3A%2F%2Fpicture1.jpg&w=384&q=75"
        srcset="/_next/image?url=http%3A%2F%2Fpicture1.jpg&w=256&q=75 1x, /_next/image?url=http%3A%2F%2Fpicture1.jpg&w=384&q=75 2x"
        style="color: transparent;"
        width="180"
      />
      <span
        class="item-metadata"
      >
        <p>
          $ 100 
          <img
            alt="free shipping icon"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=32&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 1x, /_next/image?url=%2Fimg.jpg&w=32&q=75 2x"
            style="color: transparent;"
            width="16"
          />
        </p>
        <h2>
          item1
        </h2>
      </span>
      <p
        class="item-seller"
      >
        seller1
      </p>
    </a>
    <a
      class="item-card"
      href="/items/2"
    >
      <img
        alt="picture-item-2"
        class="item-picture"
        data-nimg="1"
        decoding="async"
        height="180"
        loading="lazy"
        src="/_next/image?url=http%3A%2F%2Fpicture2.jpg&w=384&q=75"
        srcset="/_next/image?url=http%3A%2F%2Fpicture2.jpg&w=256&q=75 1x, /_next/image?url=http%3A%2F%2Fpicture2.jpg&w=384&q=75 2x"
        style="color: transparent;"
        width="180"
      />
      <span
        class="item-metadata"
      >
        <p>
          $ 200 
        </p>
        <h2>
          item2
        </h2>
      </span>
      <p
        class="item-seller"
      >
        seller2
      </p>
    </a>
  </section>
</DocumentFragment>
`;
